### PR TITLE
Promotion expiration

### PR DESCRIPTION
**Pull request by masparrow**
Resolves https://github.com/brave/brave-browser/issues/6947

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

1. Clean install
2. Join rewards, and claim a grant
3. Check your balance (17.5 BAT)
4. Close Brave
5. Edit the promotion expiresAt in the database `publisher_info_db` to be in the past (i.e. 946684800 for 00:00:00 01/01/2000)
6. Launch Brave
7. Check your balance (should now be 0.0 BAT)
8. There should be no details button (to show more grants) in wallet panel

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/brave_rewards/browser/database/publisher_info_database.cc [hunk 1]
```diff
@@ -1239,6 +1239,20 @@ bool PublisherInfoDatabase::DeleteUnblindedToken(
...
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  bool initialized = Init();
+  DCHECK(initialized);
+
+  if (!initialized) {
+    return false;
+  }
+
+  return DatabaseUnblindedToken::DeleteRecordsForPromotion(&GetDB(), promotion_id);
```
Comment (masparrow) [inline components/brave_rewards/browser/database/publisher_info_database.cc]: Lint issue - have resolved locally and will push once all PR comments are addressed

File: components/brave_rewards/browser/database/database_unblinded_token.cc [hunk 2]
```diff
@@ -16,6 +16,13 @@
 
 namespace brave_rewards {
 
+namespace  {
+
+const char* table_name_ = "unblinded_tokens";
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/database/database_unblinded_token.cc]: why are these two lines needed as we have it already defined in h file

File: components/brave_rewards/browser/database/database_unblinded_token.h [hunk 3]
```diff
@@ -33,9 +33,9 @@ class DatabaseUnblindedToken: public DatabaseTable {
 
   bool DeleteRecord(sql::Database* db, const std::vector<std::string>& id_list);
 
- private:
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/database/database_unblinded_token.h]: why was this moved?

File: components/brave_rewards/browser/rewards_service_impl.h [hunk 4]
```diff
@@ -728,6 +728,10 @@ class RewardsServiceImpl : public RewardsService,
     const std::vector<std::string>& id_list,
     ledger::ResultCallback callback) override;
 
+  void DeleteUnblindedTokensForPromotion(
+    const std::string& id_list,
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/rewards_service_impl.h]: should be `promotion_id`

File: vendor/bat-native-ledger/src/bat/ledger/internal/promotion.cc [hunk 5]
```diff
@@ -40,6 +40,34 @@ using challenge_bypass_ristretto::UnblindedToken;
 
 namespace braveledger_promotion {
 
+namespace {
+
+void HandleExpiredPromotions(
+    bat_ledger::LedgerImpl* ledger,
+    ledger::PromotionMap* promotions) {
+  // TODO(https://github.com/brave/brave-browser/issues/6363): Filtering logic
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/promotion.cc]: not sure about this TODO. It links to move db into ledger. How are they related

File: components/brave_rewards/browser/database/database_unblinded_token.cc (hunk 2 above)
Comment (masparrow) [reply components/brave_rewards/browser/database/database_unblinded_token.cc]: They were defined as class instance members, so they got allocated for every 'DatabaseUnblindedToken' instance. 
As they should be 'static' - Chromium standards say to put them in the .cc 
Also, I needed to reference 'table_name_' in my new static class method.

File: components/brave_rewards/browser/database/database_unblinded_token.h (hunk 3 above)
Comment (masparrow) [reply components/brave_rewards/browser/database/database_unblinded_token.h]: See above.

File: components/brave_rewards/browser/rewards_service_impl.h (hunk 4 above)
Comment (masparrow) [reply components/brave_rewards/browser/rewards_service_impl.h]: Wow - that's a bad oversight - good spot, thank you.

File: vendor/bat-native-ledger/src/bat/ledger/internal/promotion.cc (hunk 5 above)
Comment (masparrow) [reply vendor/bat-native-ledger/src/bat/ledger/internal/promotion.cc]: On speaking with @tmancey regarding refactoring the database, this kind of operation should be a query executed within the DB, not in code. We could actually change to this now, but I've no oversight on the DB refactor at the minute. What do you think?

File: components/brave_rewards/browser/database/publisher_info_database.cc (hunk 1 above)
Comment (masparrow) [reply components/brave_rewards/browser/database/publisher_info_database.cc]: Done.

File: components/brave_rewards/browser/rewards_service_impl.h (hunk 4 above)
Comment (masparrow) [reply components/brave_rewards/browser/rewards_service_impl.h]: Fixed.

File: components/brave_rewards/browser/database/database_unblinded_token.cc (hunk 2 above)
Comment (NejcZdovc) [reply components/brave_rewards/browser/database/database_unblinded_token.cc]: based on the fact that we do this in all other DB files, can you revert here and change them all in follow up?

Comment (masparrow) [reply components/brave_rewards/browser/database/database_unblinded_token.cc]: Sure, I don't mind - though the change was instigated by needing `table_name_` in a static method, so that would force other changes (none of them monumental). 

I could revert this and duplicate the `table_name_` to get us by - but this is taking us down the wrong route IMHO and I'd rather move the right way. But, your call, I'm happy either way.

Comment (NejcZdovc) [reply components/brave_rewards/browser/database/database_unblinded_token.cc]: sounds good, let's keep it like this for now. Please log an issue to migrate all other db files to this new approach

File: components/brave_rewards/browser/database/database_unblinded_token.cc [hunk 6]
```diff
@@ -154,4 +161,21 @@ bool DatabaseUnblindedToken::DeleteRecord(
   return statement.Run();
 }
 
+bool DatabaseUnblindedToken::DeleteRecordsForPromotion(
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/database/database_unblinded_token.cc]: nit: if function is static, we normally add `// static` above function

File: components/brave_rewards/browser/database/publisher_info_database.cc [hunk 7]
```diff
@@ -1239,6 +1239,21 @@ bool PublisherInfoDatabase::DeleteUnblindedToken(
...
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  bool initialized = Init();
+  DCHECK(initialized);
+
+  if (!initialized) {
+    return false;
+  }
+
+  return DatabaseUnblindedToken::DeleteRecordsForPromotion(&GetDB(),
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/database/publisher_info_database.cc]: nit: can you please put each param in separate like

```
return DatabaseUnblindedToken::DeleteRecordsForPromotion(
       &GetDB(),
       promotion_id);
```

File: components/services/bat_ledger/public/cpp/ledger_client_mojo_proxy.h [hunk 8]
```diff
@@ -249,6 +249,10 @@ class LedgerClientMojoProxy : public mojom::BatLedgerClient,
       const std::vector<std::string>& id_list,
       DeleteUnblindedTokenCallback callback) override;
 
+  void DeleteUnblindedTokensForPromotion(
+    const std::string& promotion_id,
```
Comment (NejcZdovc) [inline components/services/bat_ledger/public/cpp/ledger_client_mojo_proxy.h]: should be 4 spaces

File: components/services/bat_ledger/public/cpp/ledger_client_mojo_proxy.h [hunk 9]
```diff
@@ -458,6 +462,10 @@ class LedgerClientMojoProxy : public mojom::BatLedgerClient,
     CallbackHolder<DeleteUnblindedTokenCallback>* holder,
     const ledger::Result result);
 
+  static void OnDeleteUnblindedTokensForPromotion(
+    CallbackHolder<DeleteUnblindedTokensForPromotionCallback>* holder,
```
Comment (NejcZdovc) [inline components/services/bat_ledger/public/cpp/ledger_client_mojo_proxy.h]: should be 4 spaces

File: vendor/bat-native-ledger/include/bat/ledger/ledger_client.h [hunk 10]
```diff
@@ -317,6 +317,10 @@ class LEDGER_EXPORT LedgerClient {
     const std::vector<std::string>& id_list,
     ledger::ResultCallback callback) = 0;
 
+  virtual void DeleteUnblindedTokensForPromotion(
+    const std::string& promotion_id,
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/include/bat/ledger/ledger_client.h]: should be 2 more spaces

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.h [hunk 11]
```diff
@@ -596,6 +596,10 @@ class LedgerImpl : public ledger::Ledger,
     const std::vector<std::string>& id_list,
     ledger::ResultCallback callback);
 
+  void DeleteUnblindedTokensForPromotion(
+    const std::string& promotion_id,
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.h]: should be 2 more spaces

File: vendor/bat-native-ledger/src/bat/ledger/internal/promotion.cc (hunk 5 above)
Comment (NejcZdovc) [reply vendor/bat-native-ledger/src/bat/ledger/internal/promotion.cc]: personally I would remove this TODO and a reference to the issue as it will not fix in that issue

**Comment by NejcZdovc** [comment]
can you please add browser test for it?

File: components/services/bat_ledger/public/cpp/ledger_client_mojo_proxy.h (hunk 8 above)
Comment (masparrow) [reply components/services/bat_ledger/public/cpp/ledger_client_mojo_proxy.h]: Ah yes, sorry - I'd subconsciously followed the existing code - will fix.

File: vendor/bat-native-ledger/src/bat/ledger/internal/promotion.cc (hunk 5 above)
Comment (masparrow) [reply vendor/bat-native-ledger/src/bat/ledger/internal/promotion.cc]: Done

File: components/brave_rewards/browser/database/database_unblinded_token.cc (hunk 6 above)
Comment (masparrow) [reply components/brave_rewards/browser/database/database_unblinded_token.cc]: Done

File: components/brave_rewards/browser/database/publisher_info_database.cc (hunk 7 above)
Comment (masparrow) [reply components/brave_rewards/browser/database/publisher_info_database.cc]: Done

File: components/services/bat_ledger/public/cpp/ledger_client_mojo_proxy.h (hunk 8 above)
Comment (masparrow) [reply components/services/bat_ledger/public/cpp/ledger_client_mojo_proxy.h]: Done

File: components/services/bat_ledger/public/cpp/ledger_client_mojo_proxy.h (hunk 9 above)
Comment (masparrow) [reply components/services/bat_ledger/public/cpp/ledger_client_mojo_proxy.h]: Done

File: vendor/bat-native-ledger/include/bat/ledger/ledger_client.h (hunk 10 above)
Comment (masparrow) [reply vendor/bat-native-ledger/include/bat/ledger/ledger_client.h]: Done

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.h (hunk 11 above)
Comment (masparrow) [reply vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.h]: Done

**Comment by masparrow** [comment]
> can you please add browser test for it?
 Will do, and I'll investigate the promotions thing - I spotted that but thought it was correct as it was a history of the claim. I'm guessing data is cached elsewhere.

File: vendor/bat-native-ledger/include/bat/ledger/ledger_client.h [hunk 12]
```diff
@@ -311,8 +311,12 @@ class LEDGER_EXPORT LedgerClient {
     ledger::GetAllUnblindedTokensCallback callback) = 0;
 
   virtual void DeleteUnblindedTokens(
-    const std::vector<std::string>& id_list,
-    ledger::ResultCallback callback) = 0;
+      const std::vector<std::string>& id_list,
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/include/bat/ledger/ledger_client.h]: spacing was fixed for this function

File: vendor/brave-ios/Ledger/BATBraveLedger.mm [hunk 13]
```diff
@@ -1997,4 +1997,9 @@ - (void)unblindedTokensReady
   }
 }
 
+- (void)deleteUnblindedTokensForPromotion:(const std::string&)promotion_id callback:(ledger::ResultCallback)callback
+{
+  // TODO please implement
```
Comment (NejcZdovc) [inline vendor/brave-ios/Ledger/BATBraveLedger.mm]: @kylehickinson can you please add implementation of it, you can find reference for it here https://github.com/brave/brave-core/pull/4035/files#diff-804d4c26e5326ea7739bb961bd6d5dfcR168

Comment (kylehickinson) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm]: @NejcZdovc Is there a reason to do this? iOS build on master branch already does not function correctly as we are waiting on #4087

Comment (NejcZdovc) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm]: that's true, we will migrate to ledger anyway. Are you happy with just this comment for now?

Comment (kylehickinson) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm]: Yep, this is fine :)

**Comment by NejcZdovc** [comment]
CI failed on `audit-network`, not related.

File: components/brave_rewards/resources/android_page/components/pageWallet.tsx [hunk 14]
```diff
@@ -56,7 +56,8 @@ class PageWallet extends React.Component<Props, State> {
     }
 
     let claimedPromotions = promotions.filter((promotion: Rewards.Promotion) => {
-      return promotion.status === 4 // PromotionStatus::FINISHED
+      return promotion.status === 4 || // PromotionStatus::FINISHED
+        promotion.status === 5 // PromotionStatus::OVER
```
Comment (tmancey) [inline components/brave_rewards/resources/android_page/components/pageWallet.tsx]: nit: @NejcZdovc not for this PR, but `::OVER` would be better named `EXPIRED`

File: components/brave_rewards/resources/extension/brave_rewards/utils.ts [hunk 15]
```diff
@@ -41,7 +41,8 @@ export const generatePromotions = (promotions?: RewardsExtension.Promotion[]) =>
   }
 
   let claimedPromotions = promotions.filter((promotion: Rewards.Promotion) => {
-    return promotion.status === 4 // PromotionStatus::FINISHED
+    return promotion.status === 4 || // PromotionStatus::FINISHED
+      promotion.status === 5 // PromotionStatus::OVER
```
Comment (tmancey) [inline components/brave_rewards/resources/extension/brave_rewards/utils.ts]: nit: @NejcZdovc not for this PR as unrelated, but `::OVER` would be better named `EXPIRED`

File: components/brave_rewards/resources/page/components/pageWallet.tsx [hunk 16]
```diff
@@ -159,7 +159,8 @@ class PageWallet extends React.Component<Props, State> {
     }
 
     let claimedPromotions = promotions.filter((promotion: Rewards.Promotion) => {
-      return promotion.status === 4 // PromotionStatus::FINISHED
+      return promotion.status === 4 || // PromotionStatus::FINISHED
+        promotion.status === 5 // PromotionStatus::OVER
```
Comment (tmancey) [inline components/brave_rewards/resources/page/components/pageWallet.tsx]: nit: @NejcZdovc not for this PR as unrelated, but `::OVER` would be better named `EXPIRED`

File: components/brave_rewards/resources/android_page/components/pageWallet.tsx (hunk 14 above)
Comment (tmancey) [reply components/brave_rewards/resources/android_page/components/pageWallet.tsx]: @NejcZdovc Business logic like this should be moved to the libs (but not for this ticket as unrelated)

File: components/brave_rewards/resources/extension/brave_rewards/utils.ts (hunk 15 above)
Comment (tmancey) [reply components/brave_rewards/resources/extension/brave_rewards/utils.ts]: @NejcZdovc Business logic like this should be moved to the libs (but not for this ticket as unrelated)

File: components/brave_rewards/resources/page/components/pageWallet.tsx (hunk 16 above)
Comment (tmancey) [reply components/brave_rewards/resources/page/components/pageWallet.tsx]: @NejcZdovc Business logic like this should be moved to the libs (but not for this ticket as unrelated)

File: components/services/bat_ledger/public/cpp/ledger_client_mojo_proxy.cc [hunk 17]
```diff
@@ -1179,6 +1179,30 @@ void LedgerClientMojoProxy::DeleteUnblindedTokens(
...
 
+// static
+void LedgerClientMojoProxy::OnDeleteUnblindedTokensForPromotion(
+    CallbackHolder<DeleteUnblindedTokensForPromotionCallback>* holder,
+    const ledger::Result result) {
+  DCHECK(holder);
+  if (holder->is_valid()) {
+    std::move(holder->get()).Run(result);
+  }
+  delete holder;
```
Comment (tmancey) [inline components/services/bat_ledger/public/cpp/ledger_client_mojo_proxy.cc]: legacy code but if `holder` is a `nullptr` release builds will crash

File: components/brave_rewards/resources/android_page/components/pageWallet.tsx (hunk 14 above)
Comment (NejcZdovc) [reply components/brave_rewards/resources/android_page/components/pageWallet.tsx]: it's not a business logic. We send promotions to the UI and then UI decides what to show where. So for example from the same list we show notifications about new grants and list of already claimed grants

File: components/brave_rewards/resources/extension/brave_rewards/utils.ts (hunk 15 above)
Comment (NejcZdovc) [reply components/brave_rewards/resources/extension/brave_rewards/utils.ts]: see comment above

File: components/brave_rewards/resources/page/components/pageWallet.tsx (hunk 16 above)
Comment (NejcZdovc) [reply components/brave_rewards/resources/page/components/pageWallet.tsx]: see comment above

File: components/brave_rewards/resources/android_page/components/pageWallet.tsx (hunk 14 above)
Comment (NejcZdovc) [reply components/brave_rewards/resources/android_page/components/pageWallet.tsx]: cc @kylehickinson for visibility
